src/encoding/resolve.rs

```diff
@@ -0,0 +1,468 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Also licensed under MIT license, at your choice.
+
+use std::ops::Range;
+
+use bytemuck::{Pod, Zeroable};
+use moscato::pinot::FontRef;
+
+use super::{
+    glyph_cache::{CachedRange, GlyphCache, GlyphKey},
+    ramp_cache::{RampCache, Ramps},
+    DrawTag, Encoding, PathTag, StreamOffsets, Transform,
+};
+use crate::glyph::GlyphContext;
+use crate::shaders;
+
+/// Layout of a packed encoding.
+#[derive(Clone, Copy, Debug, Default, Zeroable, Pod)]
+#[repr(C)]
+pub struct Layout {
+    /// Number of draw objects.
+    pub n_draw_objects: u32,
+    /// Number of paths.
+    pub n_paths: u32,
+    /// Number of clips.
+    pub n_clips: u32,
+    /// Start of binning data.
+    pub bin_data_start: u32,
+    /// Start of path tag stream.
+    pub path_tag_base: u32,
+    /// Start of path data stream.
+    pub path_data_base: u32,
+    /// Start of draw tag stream.
+    pub draw_tag_base: u32,
+    /// Start of draw data stream.
+    pub draw_data_base: u32,
+    /// Start of transform stream.
+    pub transform_base: u32,
+    /// Start of linewidth stream.
+    pub linewidth_base: u32,
+}
+
+impl Layout {
+    /// Creates a zeroed layout.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns the path tag stream.
+    pub fn path_tags<'a>(&self, data: &'a [u8]) -> &'a [PathTag] {
+        let start = self.path_tag_base as usize * 4;
+        let end = self.path_data_base as usize * 4;
+        bytemuck::cast_slice(&data[start..end])
+    }
+
+    /// Returns the path tag stream in chunks of 4.
+    pub fn path_tags_chunked<'a>(&self, data: &'a [u8]) -> &'a [u32] {
+        let start = self.path_tag_base as usize * 4;
+        let end = self.path_data_base as usize * 4;
+        bytemuck::cast_slice(&data[start..end])
+    }
+
+    /// Returns the path data stream.
```

Comment:
This feels like it needs a caveat, it's only valid if all the paths encoded are in f32 format.

Reply:
Oops. Good catch. I just changed this to return bytes as that's the best we can do without further context. I guess we'll need to pair this up with path tags to actually extract useful data.